src/visualize/TensorProductVoronoiHandler.cpp

```diff
@@ -0,0 +1,146 @@
+{
+    std::vector<double> pointCoordinates1D;
+    for (int id = 0; id <= numCellsPerDirection; id++)
+        pointCoordinates1D.push_back(-1.0 + 2.0 * id / numCellsPerDirection);
+
+    for (int i = 0; i < std::pow(numCellsPerDirection + 1, TDim); i++)
+    {
+        Eigen::VectorXd coordinate(TDim);
+        int power = 1;
+        for (int dim = 0; dim < TDim; dim++)
+        {
+            int index = (i / power) % (numCellsPerDirection + 1);
+            coordinate(dim) = pointCoordinates1D[index];
+            power *= (numCellsPerDirection + 1);
+        }
+        mPointCoordinates.push_back(coordinate);
+    }
+
+    SetCellType();
+
+    SetVoronoiCells(numCellsPerDirection);
+}
+
+namespace NuTo
+{
+namespace Visualize
+{
+
+template <>
+void TensorProductVoronoiHandler<1>::SetCellType()
+{
+    mCellType = NuTo::eCellTypes::LINE;
+}
+
+template <>
+void TensorProductVoronoiHandler<2>::SetCellType()
+{
+    mCellType = NuTo::eCellTypes::QUAD;
+}
+
+template <>
+void TensorProductVoronoiHandler<3>::SetCellType()
+{
+    mCellType = NuTo::eCellTypes::HEXAHEDRON;
+}
+
+template <>
+void TensorProductVoronoiHandler<1>::SetVoronoiCells(int numCellsPerDirection)
+{
+    for (int i = 0; i < numCellsPerDirection; i++)
+    {
+        mVoronoiCellCorners.push_back({i, i + 1});
+    }
+}
+
+template <>
+void TensorProductVoronoiHandler<2>::SetVoronoiCells(int numCellsPerDirection)
+{
+    for (int row = 0; row < numCellsPerDirection; row++)
+    {
+        for (int col = 0; col < numCellsPerDirection; col++)
+        {
+            int start = row * (numCellsPerDirection + 1) + col;
+            mVoronoiCellCorners.push_back(
+                    {start, start + 1, start + numCellsPerDirection + 2, start + numCellsPerDirection + 1});
+        }
+    }
+}
+
+template <>
+void TensorProductVoronoiHandler<3>::SetVoronoiCells(int numCellsPerDirection)
+{
+    for (int height = 0; height < numCellsPerDirection; height++)
+    {
+        for (int row = 0; row < numCellsPerDirection; row++)
+        {
+            for (int col = 0; col < numCellsPerDirection; col++)
+            {
+                int start1 = row * (numCellsPerDirection + 1) + col +
+                             height * ((numCellsPerDirection + 1) * (numCellsPerDirection + 1));
+                int start2 = row * (numCellsPerDirection + 1) + col +
+                             (height + 1) * ((numCellsPerDirection + 1) * (numCellsPerDirection + 1));
+                mVoronoiCellCorners.push_back({start1, start1 + 1, start1 + numCellsPerDirection + 2,
+                                               start1 + numCellsPerDirection + 1, start2, start2 + 1,
+                                               start2 + numCellsPerDirection + 2, start2 + numCellsPerDirection + 1});
+            }
+        }
+    }
+}
+
+template <int TDim>
+std::vector<int> TensorProductVoronoiHandler<TDim>::WriteGeometry(const CellInterface& cell, UnstructuredGrid* grid)
```

Comment:
I claim that the methods `WriteGeometry, WriteDofValues, CellData` will be the same for any other `VoronoiHandler`. Following YAGNI, it may be fine for now. To avoid code duplication, we could have a `VoronoiHandler` base class with said methods and let classes like `TensorProduct..` derive from that. 
But I suggest a `VoronoiHandler` that works on a `VoronoiTriangulation`. This way, the creation of this geometry is separated from its visualization. 
Btw: `WriteGeometry, WriteDofValues, CellData` do not need the template argument `TDim`. For me, this is a strong indicator that we can/should split this class into a `TDim` dependent part (the actual voronoi cell generation) and a general part (the `VoronoiHandler`. Even if it is just so save compile time.
What do you think?